### ISSUE TITLE
how to impove predict result?

### ISSUE THREAD

**Issue by liaicheng** [HadoopIt/rnn-nlu#2]
HI, i used your code and train a model.  When predict with my test data, intent result seems good, but tagging task seems worse comparing to  other mate.
I changed some flags parameters, such as double 'batch_size',double 'word_embedding_size', double 'max_training_steps',double 'num_layers',no use. Can you give me other tip? :)

thanks!

**Comment by HadoopIt** [comment]
Thanks for trying out this sample code. When doing model tuning, typically I will first look at the training curve and validation curve, to see whether the model overfits or underfits. From there, I will tune accordingly to address the problem. 

Moreover, the sample code published here does not include the slot label dependency modeling as mentioned in the code comment. If there are strong dependencies among your labels/tags, connecting label/tag output back to RNN hidden state might help.

**Comment by liaicheng** [comment]
The training like below:
![image](https://cloud.githubusercontent.com/assets/3361572/20829751/662145fc-b8b8-11e6-85f0-fd27a65b5f59.png)

It seems that test fl score and accuracy is good. 
I didn't get your meaning:

> Moreover, the sample code published here does not include the slot label dependency modeling as mentioned in the code comment. If there are strong dependencies among your labels/tags, connecting label/tag output back to RNN hidden state might help.

PS: I don't know if you know chinese, if you can , may i leave messeage in chinese ?:)

**Comment by hariom-yadaw** [comment]
I have some questions as below regarding the data set and implementation on github repo as below.

1. Since we create vocabulary only from training input seq(train.seq.in), the new names for the 2 slots(from_City & to_City), will be taken as _UNK in vocabulary. Having said this, will RNN be able to detect this slots correctly?

2. If YES, then how actually it does so to detect slots correctly without having the word in vocab build from training data ? Whats the logic behind it?

3. I trained on the data set in your github repo, I was not able to get good the f1 score or accuracy. What are the most effective ways to improve it. 


If you have free time, I request you to please answer my questions above. I want to learn about these. Thanks for your time and concern.

**Comment by hariom-yadaw** [comment]
@liaicheng  To get the F1 score calculated properly from conlleval.pl, do we need to have some fixed format for labels ? Because I didn't have the same format in dataset labels, and somehow its printing F1 scores as zero(but actually F1 score is not zero when I manually check the test_results)

**Comment by hariom-yadaw** [comment]
@HadoopIt @liaicheng Have anyone tried using pre-trained embedding word vector to improve the accuracy ? how to do this and where to add this embedding in this project ? Thanks!

**Comment by liaicheng** [comment]
@hariom-yadaw actually i process the slot data as the sample show, so i can get the f1 score.
i didn't make pre-trained,,accuracy seems not very well. i think you can try to change training parameters and do it again, maybe something would be better.

**Comment by HadoopIt** [comment]
@hariom-yadaw Slot labels for words that do not appear in training set might be inferred from the structure in the sequence, e.g. structure like "flight from A to B". 

Pre-trained word embedding can be fed to the graph when you do tf.session.run, just like feeding other input values. More details:
http://stackoverflow.com/questions/35687678/using-a-pre-trained-word-embedding-word2vec-or-glove-in-tensorflow

**Comment by hariom-yadaw** [comment]
Thanks a lot @HadoopIt In case we use pre-trained embedding, do we need to change other things in code like vocab will be taken from pre-trained file and not from train.seq.in ?

**Comment by hariom-yadaw** [comment]
Also I get weird results as below when I run it on my dataset. the accuracy is 100% from 1st step itself because it checking _UNK against _UNK. I printed `print(ref_label_list)` and `print(hyp_label_list)` while testing on Test and Valid dataset. 

Reading train/valid/test data (training set limit: 0).
global step 300 step-time 0.08. Training perplexity 1.53
  Eval accuracy: 100.00 21/21
  Eval f1-score: 0.00
['_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK']
['_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK']
  Test accuracy: 100.00 18/18
  Test f1-score: 0.00
['_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK']
['_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK', '_UNK']


My data set.
------------------------------
test.seq.in
............................................
can you give me a ppt on Plantation ?
I want your help on Higher Education.
I expect you to help me on Education System.
Show me some data on Mahabharat.
I need your help on Moon.
what can you show me on Sun ?
I would like to have your views on Water Pollution.
What ideas do you have on Traffic System.
Hey
Hey, what is your name ?
you can call me Pankaj.
You can have my name as Gaurav.
what help you can provide ?
what you are good at ?
Entertain me.
Tell me a joke.
Hi!
Thanks for your great help.


test.seq.out(slot tagging)
..............................................................
0 0 0 0 0 0 0 B_eduTopic 0
0 0 0 0 0 B_eduTopic E_eduTopic 0
0 0 0 0 0 0 0 B_eduTopic E_eduTopic 0
0 0 0 0 0 B_eduTopic 0
0 0 0 0 0 B_eduTopic 0
0 0 0 0 0 0 B_eduTopic 0
0 0 0 0 0 0 0 0 B_eduTopic E_eduTopic 0
0 0 0 0 0 0 B_eduTopic E_eduTopic 0
0
0 0 0 0 0 0 0
0 0 0 0 B_user.name 0
0 0 0 0 0 0 B_user.name 0
0 0 0 0 0 0
0 0 0 0 0 0
0 0 0
0 0 0 0 0
0 0
0 0 0 0 0 0
0 0 0 0
0 0 0
0

test.label(intent)
................................................
edu.Topic
edu.Topic
edu.Topic
edu.Topic
edu.Topic
edu.Topic
edu.Topic
edu.Topic
greet.welcome
name.ask
name.tell
name.tell
help.type
help.type
mood.laugh
mood.laugh
greet.welcome
thanks.bye
thanks.bye
thnaks.bye
thanks.bye
